pkg/resource/deploy/source_eval.go

```diff
@@ -196,26 +202,143 @@ func (iter *evalSourceIterator) forkRun(opts Options) {
 	}()
 }
 
+type defaultProviderResponse struct {
+	ref providers.Reference
+	err error
+}
+
+type defaultProviderRequest struct {
+	pkg      tokens.Package
+	response chan<- defaultProviderResponse
+}
+
+type defaultProviders struct {
+	versions  map[tokens.Package]*semver.Version
+	providers map[tokens.Package]providers.Reference
+	config    plugin.ConfigSource
+
+	requests chan defaultProviderRequest
+	regChan  chan<- RegisterResourceEvent
+	cancel   <-chan bool
+}
+
+func (d *defaultProviders) newRegisterDefaultProviderEvent(
+	pkg tokens.Package) (RegisterResourceEvent, <-chan *RegisterResult, error) {
+
+	// Attempt to get the config for the package.
+	cfg, err := d.config.GetPackageConfig(pkg)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create the inputs for the provider resource.
+	inputs := make(resource.PropertyMap)
+	for k, v := range cfg {
+		inputs[resource.PropertyKey(k.Name())] = resource.NewStringProperty(v)
+	}
+	if version := d.versions[pkg]; version != nil {
+		inputs["version"] = resource.NewStringProperty(version.String())
+	}
+
+	// Create the result channel and the event.
+	done := make(chan *RegisterResult)
+	event := &registerResourceEvent{
+		goal: resource.NewGoal(providers.MakeProviderType(pkg), "default", true, inputs, "", false, nil, ""),
+		done: done,
+	}
+	return event, done, nil
+}
+
+func (d *defaultProviders) serve() {
+	for req := range d.requests {
+		logging.V(5).Infof("handling default provider request for package %s", req.pkg)
+
+		ref, ok := d.providers[req.pkg]
+		if !ok {
+			logging.V(5).Infof("registering default provider for package %s", req.pkg)
+
+			event, done, err := d.newRegisterDefaultProviderEvent(req.pkg)
+			if err != nil {
+				req.response <- defaultProviderResponse{err: err}
+				continue
+			}
+
+			d.regChan <- event
+
+			logging.V(5).Infof("waiting for default provider for package %s", req.pkg)
+
+			var result *RegisterResult
+			select {
+			case result = <-done:
+			case <-d.cancel:
+				return
```

Comment:
I don't think returning is the right thing to do here when we're canceled, we're going to leave anybody blocked in `getDefaultProviderRef` deadlocked waiting for the response. We should at least be closing the response channel.